fs/metrics.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+#include "rocksdb/env.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+struct ZenFSMetrics {
+public:
+  typedef uint32_t Label;
+  typedef uint32_t ReporterType;
+  // We give an enum to identify the reporters and an enum to identify the reporter types: 
+  // ZenFSMetricsHistograms and ZenFSMetricsReporterType, respectively, at the end of the code.
+public:
+  ZenFSMetrics() {}
+  virtual ~ZenFSMetrics() {}
+public:
+  // Add a reporter named label.
+  // You can give a type for type-checking.
+  virtual void AddReporter(Label label, ReporterType type = 0) = 0;
+  // Report a value for the reporter named label.
+  // You can give a type for type-checking.
+  virtual void Report(Label label, size_t value, ReporterType type_check = 0) = 0;
+public: 
+  // Syntactic sugars for type-checking.
+  // Overwrite them if you think type-checking is necessary.
+  virtual void ReportQPS(Label label, size_t qps) { Report(label, qps, 0); }
+  virtual void ReportThroughput(Label label, size_t throughput) { Report(label, throughput, 0); }
+  virtual void ReportLatency(Label label, size_t latency) { Report(label, latency, 0); }
+  virtual void ReportGeneral(Label label, size_t data) { Report(label, data, 0); }
+  // and more
+};
+
+struct NoZenFSMetrics : public ZenFSMetrics {
```

Comment:
A default implementation (in addition to `NoZenFSMetrics`) that actually records and reports metrics would be great. It would make the code a lot more useful, and would be a nice example of how to implement custom metric collection.

Reply:
> 

I've given a default implementation as a sample, so hopefully this will help.

Reply:
Great, thanks! I think we could eventually make a config item to enable the default reporter.